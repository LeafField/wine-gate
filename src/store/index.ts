import { create } from "zustand";

type User = {
  username: string;
  id: string;
  email: string | undefined;
};

type Menu = "close" | "member" | "free" | "category";

type Store = {
  user: User | null;
  setUser: (user: User) => void;
  logout: () => void;
  menu: Menu;
  setMenu: (menu: Menu) => void;
};

export const useStore = create<Store>((set) => ({
  user: null,
  setUser: (user) => set({ user }),
  logout: () => set({ user: null }),
  menu: "close",
  setMenu: (menu) => set({ menu }),
}));
