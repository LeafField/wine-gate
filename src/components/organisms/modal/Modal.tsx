"use client";
import React, { FC, useEffect, useRef } from "react";
import { useStore } from "../../../store";

const Modal: FC = () => {
  const modalRef = useRef<HTMLDialogElement>(null);
  const { modal, clearModal } = useStore();

  useEffect(() => {
    if (modal.length > 0) {
      modalRef.current?.showModal();
    } else {
      modalRef.current?.close();
    }
  }, [modal]);

  return (
    <dialog
      ref={modalRef}
      className="relative animate-fadeIn border border-gray bg-white px-4 py-4 shadow-md backdrop:bg-[rgba(255,255,255,0.6)] backdrop:backdrop-blur-[4px] max-tablet:mx-4 tablet:w-[400px]"
    >
      {modal.map((message, index) => (
        <p key={index} className={`text-base`}>
          {message}
        </p>
      ))}
      <button
        type="button"
        onClick={clearModal}
        className="mx-auto mt-4 block border-2 border-gray px-3 py-1 transition-colors duration-300 hover:border-blue focus:border-blue focus:outline-none"
      >
        閉じる
      </button>
    </dialog>
  );
};

export default Modal;
