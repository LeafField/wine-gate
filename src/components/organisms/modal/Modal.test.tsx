import { screen, render, renderHook, act } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Modal from "./Modal";
import { useStore } from "../../../store";

describe("Modalの単体テスト", () => {
  test("初期状態では、何も表示されない", () => {
    render(<Modal />);
    expect(screen.queryByRole("dialog")).toBeNull();
  });

  test("setModalを実行すると、dialogが表示される", () => {
    const { result } = renderHook(() => useStore());
    act(() => {
      result.current.setModal(["メッセージ1", "メッセージ2"]);
    });
    render(<Modal />);
    expect(screen.getByRole("dialog")).toBeInTheDocument();
    expect(screen.getByText("メッセージ1")).toBeInTheDocument();
    expect(screen.getByText("メッセージ2")).toBeInTheDocument();
  });

  test("閉じるボタンを押すと、dialogが非表示になる", async () => {
    render(<Modal />);
    await userEvent.click(screen.getByRole("button"));
    expect(screen.queryByRole("dialog")).toBeNull();
  });

  test("スナップショットテスト", () => {
    const { result } = renderHook(() => useStore());
    act(() => {
      result.current.setModal([
        "ご登録いただいたメールアドレスへ確認メールを送信しました。",
        "メール内のリンクより登録を完了してください。",
      ]);
    });

    const { container } = render(<Modal />);
    expect(container).toMatchSnapshot();
  });
});
