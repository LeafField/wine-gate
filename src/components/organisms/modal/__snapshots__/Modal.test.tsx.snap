// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modalの単体テスト スナップショットテスト 1`] = `
<div>
  <dialog
    class="relative animate-fadeIn border border-gray bg-white px-4 py-4 shadow-md backdrop:bg-[rgba(255,255,255,0.6)] backdrop:backdrop-blur-[4px] max-tablet:mx-4 tablet:w-[400px]"
  >
    <p
      class="text-base"
    >
      ご登録いただいたメールアドレスへ確認メールを送信しました。
    </p>
    <p
      class="text-base"
    >
      メール内のリンクより登録を完了してください。
    </p>
    <button
      class="mx-auto mt-4 block border-2 border-gray px-3 py-1 transition-colors duration-300 hover:border-blue focus:border-blue focus:outline-none"
      type="button"
    >
      閉じる
    </button>
  </dialog>
</div>
`;
